File: java/src/com/swiftwombat/aoc/y2023/Day01.java
package com.swiftwombat.aoc.y2023;

import java.io.IOException;
import java.util.HashMap;
import java.util.Map;
import java.util.concurrent.atomic.AtomicInteger;
import java.util.regex.Pattern;

import com.swiftwombat.aoc.Day;

/**
 * Day 1: Trebuchet?!
 * 
 * @see https://adventofcode.com/2023/day/1
 * @author Zachary Cockshutt
 * @since 2023-12-03
 */
public class Day01 extends Day {

    // @formatter:off
    private static final Map<String, String> digitWords = new HashMap<>()
    {{
        put("one",  "1"); put("two",  "2"); put("three","3"); 
        put("four", "4"); put("five", "5"); put("six",  "6"); 
        put("seven","7"); put("eight","8"); put("nine", "9");
    }};
    // @formatter:on

    @Override
    public String partOne() throws IOException {
        var sum = new AtomicInteger(0);
        this.forEachInputLine(line -> sum.set(sum.get() + this.parseInt(line)));
        return sum.toString();
    }

    @Override
    public String partTwo() throws IOException {
        var sum = new AtomicInteger(0);
        Pattern p = Pattern.compile("(?=(one|t(wo|hree)|f(our|ive)|s(ix|even)|eight|nine))");
        this.forEachInputLine(line -> {
            var matcher = p.matcher(line);
            line = matcher.replaceAll(m -> digitWords.get(m.group(1)));
            sum.set(sum.get() + this.parseInt(line));
        });
        return sum.toString();
    }

    private int parseInt(String s) {
        s = s.replaceAll("\\D", "");
        s = "" + s.charAt(0) + s.charAt(s.length() - 1);
        return Integer.parseInt(s);
    }
}

File: csharp/src/y2025/Day01.cs
namespace aoc.y2025;

/**
 * Day 1: Secret Entrance
 *
 * @see https://adventofcode.com/2025/day/1
 * @author Zachary Cockshutt
 * @since  2025-12-01
 */
public class Day01 : Day
{
    public override string PartOne()
    {
        int password = 0;
        int dial = 50;
        ForEachInputLine(line =>
        {
            int delta = ParseDelta(line);
            dial = Mod(dial + delta, 100);
            if (dial == 0) { password++; }
        });
        return password.ToString();
    }

    public override string PartTwo()
    {
        int password = 0;
        int dial = 50;
        ForEachInputLine(line =>
        {
            int delta = ParseDelta(line);
            int sign = Math.Sign(delta);
            int normalisedDial = Mod(100 + dial * sign, 100);
            password += (normalisedDial + delta * sign) / 100;
            dial = Mod(dial + delta, 100);
        });
        return password.ToString();
    }

    private static int ParseDelta(string line)
    {
        char dir = line[0];
        int delta = int.Parse(line[1..]);
        return dir == 'L' ? -delta : delta;
    }

    private static int Mod(int x, int m)
    {
        int r = x % m;
        return r < 0 ? r + m : r;
    }
}
